src/db/transaction.cr

```diff
@@ -0,0 +1,118 @@
+module DB
+  abstract class Transaction
+    include Disposable
+    include BeginTransaction
+
+    abstract def connection : Connection
+
+    def commit
+      close!
+    end
+
+    def rollback
+      close!
+    end
+
+    private def close!
+      raise DB::Error.new("Transaction already closed") if closed?
+      close
+    end
+
+    abstract def release_from_nested_transaction
+  end
+
+  class TopLevelTransaction < Transaction
+    getter connection
+    # :nodoc:
+    property savepoint_name : String? = nil
+
+    def initialize(@connection : Connection)
+      @nested_transaction = false
+      @connection.perform_begin_transaction
+    end
+
+    def commit
+      @connection.perform_commit_transaction
+      super
+    end
+
+    def rollback
+      @connection.perform_rollback_transaction
+      super
+    end
+
+    protected def do_close
+      connection.release_from_transaction
+    end
+
+    def begin_transaction : Transaction
+      raise DB::Error.new("There is an existing nested transaction in this transaction") if @nested_transaction
+      @nested_transaction = true
+      create_save_point_transaction(self)
+    end
+
+    # :nodoc:
+    def create_save_point_transaction(parent : Transaction) : SavePointTransaction
+      # TODO should we wrap this in a mutex?
+      previous_savepoint = @savepoint_name
+      savepoint_name = if previous_savepoint
+                         previous_savepoint.succ
+                       else
+                         # random prefix to avoid determinism
+                         "crystal_#{Random.rand(10_000)}_00001"
```

Comment:
`1_000_000` ? 

the usage of the prefix is also to when dealing with a pool, so not all savepoints of all connections are named equally.

When the same connection is returned to the pool _it should_ have no longer outstanding transactions. dsl and the design enforce that when using the public api. So the risk is about dirty state between checkouts in the same connection. For long living connections the numbers will be reused, but again, between different connections the names can be the same without problems to the db.

We can prefix with the `connection.object_id` also...